utils/disk_space_monitor.cc

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright (C) 2024-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+#include <chrono>
+#include <seastar/core/coroutine.hh>
+#include <seastar/core/reactor.hh>
+
+#include "utils/disk_space_monitor.hh"
+#include "utils/log.hh"
+
+namespace utils {
+
+seastar::logger logger("disk_space");
+
+void disk_space_monitor::start(std::set<std::filesystem::path> data_dirs, config cfg) {
+    if (data_dirs.empty()) {
+        on_internal_error(logger, "data_dir must be non-empty");
+    }
+    if (!_data_dirs.empty()) {
+        on_internal_error(logger, "disk_space_monitor already started");
+    }
+    _data_dirs = std::move(data_dirs);
+    _cfg = std::move(cfg);
+    _timer.set_callback(cfg.sched_group, [this] { (void)poll(); });
+    (void)poll();
+}
+
+future<> disk_space_monitor::stop() noexcept {
+    _timer.cancel();
+    co_await _async_gate.close();
+}
+
+future<> disk_space_monitor::poll() {
+    auto gh = _async_gate.hold();
```

Comment:
It's better to run the entire loop in a coroutine and wait for it in stop(). Overall it tends to be simpler.

Reply:
Sounds good, thanks.